dashboard/app/assets/javascripts/client_state.js

```diff
@@ -41,16 +49,33 @@ dashboard.clientState.levelProgress = function(level) {
 };
 
 /**
+ * Tracks the users progress after they click run
+ * @param {boolean} result - Whether the user's solution is successful
+ * @param {number} lines - Number of lines of code user wrote in this solution
+ * @param {number} testResult - Indicates pass, fail, perfect
+ * @param {number} scriptLevelId - Which level this is for
+ */
+dashboard.clientState.trackProgress = function(result, lines, testResult, scriptLevelId) {
+  if (result) {
+    addLines(lines);
+  }
+
+  if (testResult > dashboard.clientState.levelProgress(scriptLevelId)) {
```

Comment:
By the way, if it's annoying to write dashboard.clientState.levelProgress, you could just define "function levelProgress" inside the closure and then assign " dashboard.clientState.levelProgress = levelProgress". Not important if you don't mind though, either way is fine.